File: solutions/python/code-carl/ch02_list/t7_lc0160_get_intersection_node.py
"""LeetCode 160. Intersection of Two Linked Lists
https://leetcode.cn/problems/intersection-of-two-linked-lists/
"""

from __future__ import annotations

from typing import Optional

from list_node import ListNode


class Solution:
    def getIntersectionNode(self, headA: Optional[ListNode], headB: Optional[ListNode]) -> Optional[ListNode]:
        ptrA, ptrB = headA, headB
        lenA, lenB = 0, 0
        while ptrA is not None:
            ptrA = ptrA.next
            lenA += 1
        while ptrB is not None:
            ptrB = ptrB.next
            lenB += 1

        ptrA, ptrB = headA, headB
        if lenB > lenA:
            lenA, lenB = lenB, lenA
            ptrA, ptrB = ptrB, ptrA

        for _ in range(lenA - lenB):
            ptrA = ptrA.next

        while ptrA is not None and ptrA != ptrB :
            ptrA = ptrA.next
            ptrB = ptrB.next
        return None if ptrA is None else ptrA


if __name__ == "__main__":
    solution = Solution()

    common = ListNode(8, ListNode(4, ListNode(5)))
    head_a = ListNode(4, ListNode(1, common))
    head_b = ListNode(5, ListNode(6, ListNode(1, common)))
    assert solution.getIntersectionNode(head_a, head_b) is common

    head_c = ListNode(2, ListNode(6, ListNode(4)))
    head_d = ListNode(1, ListNode(5))
    assert solution.getIntersectionNode(head_c, head_d) is None

    print("All tests passed.")

File: solutions/java/code-carl/ch02_list/T7_LC0160_GetIntersectionNode.java
/**
 * Practice template for:
 * - LeetCode 160. Intersection of Two Linked Lists
 *   https://leetcode.cn/problems/intersection-of-two-linked-lists/
 */
public class T7_LC0160_GetIntersectionNode {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        int lenA = 0, lenB = 0;
        ListNode ptrA = headA, ptrB = headB;

        while (ptrA != null) {
            lenA++;
            ptrA = ptrA.next;
        }
        ptrA = headA;

        while (ptrB != null) {
            lenB++;
            ptrB = ptrB.next;
        }
        ptrB = headB;

        if (lenB > lenA) {
            int tmp = lenB;
            lenB = lenA;
            lenA = tmp;
            ListNode tmpN = ptrB;
            ptrB = ptrA;
            ptrA = tmpN;
        }

        int gap = lenA - lenB;
        while (gap-- > 0) {
            ptrA = ptrA.next;
        }

        while (ptrA != null && ptrB != null) {
            if (ptrA == ptrB) {
                return ptrA;
            }
            ptrA = ptrA.next;
            ptrB = ptrB.next;
        }
        return null;
    }
}
